File: crates/core/component/shielded-pool/src/gnark/transfer_witness_binary.rs
use anyhow::{bail, Context, Result};

use crate::{
    gnark::{
        binary::encode_triple_path_32,
        binary::{encode_vec_32, put_bytes, put_u32, put_u64, put_u8, BinaryCursor},
        transfer_witness::{
            TransferChangeOutputWitnessV18, TransferComplianceCiphertextWitnessV18,
            TransferOptionalSpendWitnessV18, TransferReceiverOutputWitnessV18,
            TransferRequiredSpendWitnessV18, TransferTierRandomizersWitnessV18, TransferWitnessV18,
        },
        typed::{
            decode_indexed_leaf, encode_indexed_leaf, encode_merkle_path, encode_point_affine,
        },
    },
    transfer::TRANSFER_PROOF_LABEL,
};

const TRANSFER_WITNESS_MAGIC: &[u8; 4] = b"PTWG";
const TRANSFER_WITNESS_VERSION: u32 = 18;

impl TransferWitnessV18 {
    pub fn encode(&self) -> Result<Vec<u8>> {
        let mut buf = Vec::new();
        put_bytes(&mut buf, TRANSFER_WITNESS_MAGIC);
        put_u32(&mut buf, TRANSFER_WITNESS_VERSION);
        put_u32(&mut buf, 0);
        put_bytes(&mut buf, &self.anchor);
        put_bytes(&mut buf, &self.asset_anchor);
        put_bytes(&mut buf, &self.compliance_anchor);
        put_bytes(&mut buf, &self.target_timestamp);
        put_bytes(&mut buf, &self.claimed_statement_hash);
        for tag in &self.routing_tags {
            put_bytes(&mut buf, tag);
        }
        put_bytes(&mut buf, &self.routing_parameter_set_id);
        put_bytes(&mut buf, &self.recent_position_floor);
        put_bytes(&mut buf, &self.action_balance_blinding);
        put_bytes(&mut buf, &self.nk);
        encode_merkle_path(&mut buf, &self.asset_path)?;
        put_u64(&mut buf, self.asset_position);
        encode_indexed_leaf(&mut buf, &self.asset_indexed_leaf);
        put_u8(&mut buf, u8::from(self.is_regulated));
        put_u8(&mut buf, self.regulated_precision);
        put_u8(&mut buf, self.unregulated_precision);
        put_u64(&mut buf, self.routing_as_of_height);
        encode_merkle_path(&mut buf, &self.sender_compliance_path)?;
        put_u64(&mut buf, self.sender_compliance_position);
        put_bytes(&mut buf, &self.sender_slot_id);
        put_bytes(&mut buf, &self.sender_slot_derivation);
        put_bytes(&mut buf, &self.sender_d);
        put_bytes(&mut buf, &self.transfer_nonce_root);
        encode_vec_32(&mut buf, &self.detection_ciphertext)?;
        put_bytes(&mut buf, &self.sender_subject_derivation);
        put_bytes(&mut buf, &self.output_subject_derivation);
        put_bytes(&mut buf, &self.ring_id_hash);
        put_bytes(&mut buf, &self.policy_id_hash);
        put_bytes(&mut buf, &self.resource_hash);
        put_bytes(&mut buf, &self.permission_hash);
        put_bytes(&mut buf, &self.metadata_target_timestamp);
        put_bytes(&mut buf, &self.sender_core_salt);
        put_bytes(&mut buf, &self.sender_ext_salt);
        put_bytes(&mut buf, &self.output_core_salt);
        put_bytes(&mut buf, &self.output_ext_salt);
        encode_compliance_tier(&mut buf, &self.sender_core)?;
        encode_compliance_tier(&mut buf, &self.sender_ext)?;
        encode_compliance_tier(&mut buf, &self.output_core)?;
        encode_compliance_tier(&mut buf, &self.output_ext)?;
        encode_randomizers(&mut buf, &self.sender_randomizers);
        encode_randomizers(&mut buf, &self.output_randomizers);

        encode_required_spend(&mut buf, &self.required_spend)?;
        encode_optional_spend(&mut buf, &self.optional_spend)?;
        encode_receiver_output(&mut buf, &self.receiver_output)?;
        encode_change_output(&mut buf, &self.change_output);

        encode_point_affine(&mut buf, &self.ak_affine);
        encode_point_affine(&mut buf, &self.asset_indexed_leaf_dk_pub_affine);
        encode_point_affine(&mut buf, &self.asset_indexed_leaf_ring_pk_affine);
        encode_point_affine(&mut buf, &self.sender_diversified_generator_affine);
        encode_point_affine(&mut buf, &self.sender_transmission_key_affine);

        let total_len = u32::try_from(buf.len()).context("encoded transfer witness exceeds u32")?;
        buf[8..12].copy_from_slice(&total_len.to_le_bytes());
        Ok(buf)
    }

    pub fn decode(bytes: &[u8]) -> Result<Self> {
        let mut cursor = BinaryCursor::new(bytes);
        if cursor.read_fixed::<4>()? != *TRANSFER_WITNESS_MAGIC {
            bail!("invalid transfer witness magic");
        }
        let version = cursor.read_u32()?;
        if version != TRANSFER_WITNESS_VERSION {
            bail!("unsupported transfer witness version {version}");
        }
        let total_length = cursor.read_u32()?;
        if total_length as usize != bytes.len() {
            bail!(
                "transfer witness length mismatch: header={}, actual={}",
                total_length,
                bytes.len()
            );
        }

        let witness = Self {
            total_length,
            anchor: cursor.read_fixed::<32>()?,
            asset_anchor: cursor.read_fixed::<32>()?,
            compliance_anchor: cursor.read_fixed::<32>()?,
            target_timestamp: cursor.read_fixed::<32>()?,
            claimed_statement_hash: cursor.read_fixed::<32>()?,
            routing_tags: [cursor.read_fixed::<32>()?, cursor.read_fixed::<32>()?],
            routing_parameter_set_id: cursor.read_fixed::<32>()?,
            recent_position_floor: cursor.read_fixed::<32>()?,
            action_balance_blinding: cursor.read_fr()?,
            nk: cursor.read_fixed::<32>()?,
            asset_path: cursor.read_merkle_path()?,
            asset_position: cursor.read_u64()?,
            asset_indexed_leaf: decode_indexed_leaf(&mut cursor)?,
            is_regulated: cursor.read_bool()?,
            regulated_precision: cursor.read_u8()?,
            unregulated_precision: cursor.read_u8()?,
            routing_as_of_height: cursor.read_u64()?,
            sender_compliance_path: cursor.read_merkle_path()?,
            sender_compliance_position: cursor.read_u64()?,
            sender_slot_id: cursor.read_fixed::<32>()?,
            sender_slot_derivation: cursor.read_fixed::<32>()?,
            sender_d: cursor.read_fixed::<32>()?,
            transfer_nonce_root: cursor.read_fixed::<32>()?,
            detection_ciphertext: cursor.read_vec_32()?,
            sender_subject_derivation: cursor.read_fixed::<32>()?,
            output_subject_derivation: cursor.read_fixed::<32>()?,
            ring_id_hash: cursor.read_fixed::<32>()?,
            policy_id_hash: cursor.read_fixed::<32>()?,
            resource_hash: cursor.read_fixed::<32>()?,
            permission_hash: cursor.read_fixed::<32>()?,
            metadata_target_timestamp: cursor.read_fixed::<32>()?,
            sender_core_salt: cursor.read_fixed::<32>()?,
            sender_ext_salt: cursor.read_fixed::<32>()?,
            output_core_salt: cursor.read_fixed::<32>()?,
            output_ext_salt: cursor.read_fixed::<32>()?,
            sender_core: decode_compliance_tier(&mut cursor)?,
            sender_ext: decode_compliance_tier(&mut cursor)?,
            output_core: decode_compliance_tier(&mut cursor)?,
            output_ext: decode_compliance_tier(&mut cursor)?,
            sender_randomizers: decode_randomizers(&mut cursor)?,
            output_randomizers: decode_randomizers(&mut cursor)?,
            required_spend: decode_required_spend(&mut cursor)?,
            optional_spend: decode_optional_spend(&mut cursor)?,
            receiver_output: decode_receiver_output(&mut cursor)?,
            change_output: decode_change_output(&mut cursor)?,
            ak_affine: cursor.read_point_affine()?,
            asset_indexed_leaf_dk_pub_affine: cursor.read_point_affine()?,
            asset_indexed_leaf_ring_pk_affine: cursor.read_point_affine()?,
            sender_diversified_generator_affine: cursor.read_point_affine()?,
            sender_transmission_key_affine: cursor.read_point_affine()?,
        };

        cursor.finish(TRANSFER_PROOF_LABEL)?;
        Ok(witness)
    }
}

fn encode_compliance_tier(
    buf: &mut Vec<u8>,
    tier: &TransferComplianceCiphertextWitnessV18,
) -> Result<()> {
    put_bytes(buf, &tier.c2);
    encode_vec_32(buf, &tier.ciphertext)?;
    encode_point_affine(buf, &tier.epk_affine);
    Ok(())
}

fn decode_compliance_tier(
    cursor: &mut BinaryCursor<'_>,
) -> Result<TransferComplianceCiphertextWitnessV18> {
    Ok(TransferComplianceCiphertextWitnessV18 {
        c2: cursor.read_fixed::<32>()?,
        ciphertext: cursor.read_vec_32()?,
        epk_affine: cursor.read_point_affine()?,
    })
}

fn encode_randomizers(buf: &mut Vec<u8>, randomizers: &TransferTierRandomizersWitnessV18) {
    put_bytes(buf, &randomizers.core);
    put_bytes(buf, &randomizers.ext);
}

fn decode_randomizers(cursor: &mut BinaryCursor<'_>) -> Result<TransferTierRandomizersWitnessV18> {
    Ok(TransferTierRandomizersWitnessV18 {
        core: cursor.read_fr()?,
        ext: cursor.read_fr()?,
    })
}

fn encode_required_spend(buf: &mut Vec<u8>, spend: &TransferRequiredSpendWitnessV18) -> Result<()> {
    put_bytes(buf, &spend.nullifier);
    put_bytes(buf, &spend.spent_note_blinding);
    put_bytes(buf, &spend.spent_note_amount);
    put_bytes(buf, &spend.spent_note_asset_id);
    put_u64(buf, spend.state_commitment_position);
    encode_triple_path_32(buf, &spend.state_commitment_auth_path)?;
    put_bytes(buf, &spend.spend_auth_randomizer);
    encode_point_affine(buf, &spend.rk_affine);
    put_u8(buf, u8::from(spend.history_required));
    Ok(())
}

fn decode_required_spend(cursor: &mut BinaryCursor<'_>) -> Result<TransferRequiredSpendWitnessV18> {
    Ok(TransferRequiredSpendWitnessV18 {
        nullifier: cursor.read_fixed::<32>()?,
        spent_note_blinding: cursor.read_fixed::<32>()?,
        spent_note_amount: cursor.read_fixed::<32>()?,
        spent_note_asset_id: cursor.read_fixed::<32>()?,
        state_commitment_position: cursor.read_u64()?,
        state_commitment_auth_path: cursor.read_triple_path_32()?,
        spend_auth_randomizer: cursor.read_fr()?,
        rk_affine: cursor.read_point_affine()?,
        history_required: cursor.read_bool()?,
    })
}

fn encode_optional_spend(buf: &mut Vec<u8>, spend: &TransferOptionalSpendWitnessV18) -> Result<()> {
    put_bytes(buf, &spend.nullifier);
    put_bytes(buf, &spend.spent_note_blinding);
    put_bytes(buf, &spend.spent_note_amount);
    put_u64(buf, spend.state_commitment_position);
    encode_triple_path_32(buf, &spend.state_commitment_auth_path)?;
    put_bytes(buf, &spend.spend_auth_randomizer);
    encode_point_affine(buf, &spend.rk_affine);
    put_u8(buf, u8::from(spend.is_dummy));
    put_bytes(buf, &spend.dummy_nullifier_seed);
    put_u8(buf, u8::from(spend.history_required));
    Ok(())
}

fn decode_optional_spend(cursor: &mut BinaryCursor<'_>) -> Result<TransferOptionalSpendWitnessV18> {
    Ok(TransferOptionalSpendWitnessV18 {
        nullifier: cursor.read_fixed::<32>()?,
        spent_note_blinding: cursor.read_fixed::<32>()?,
        spent_note_amount: cursor.read_fixed::<32>()?,
        state_commitment_position: cursor.read_u64()?,
        state_commitment_auth_path: cursor.read_triple_path_32()?,
        spend_auth_randomizer: cursor.read_fr()?,
        rk_affine: cursor.read_point_affine()?,
        is_dummy: cursor.read_bool()?,
        dummy_nullifier_seed: cursor.read_fixed::<32>()?,
        history_required: cursor.read_bool()?,
    })
}

fn encode_receiver_output(
    buf: &mut Vec<u8>,
    output: &TransferReceiverOutputWitnessV18,
) -> Result<()> {
    put_bytes(buf, &output.note_commitment);
    put_bytes(buf, &output.created_note_blinding);
    put_bytes(buf, &output.created_note_amount);
    encode_merkle_path(buf, &output.recipient_compliance_path)?;
    put_u64(buf, output.recipient_compliance_position);
    put_bytes(buf, &output.recipient_slot_id);
    put_bytes(buf, &output.recipient_slot_derivation);
    put_bytes(buf, &output.recipient_d);
    encode_point_affine(buf, &output.recipient_diversified_generator_affine);
    encode_point_affine(buf, &output.recipient_transmission_key_affine);
    Ok(())
}

fn decode_receiver_output(
    cursor: &mut BinaryCursor<'_>,
) -> Result<TransferReceiverOutputWitnessV18> {
    Ok(TransferReceiverOutputWitnessV18 {
        note_commitment: cursor.read_fixed::<32>()?,
        created_note_blinding: cursor.read_fixed::<32>()?,
        created_note_amount: cursor.read_fixed::<32>()?,
        recipient_compliance_path: cursor.read_merkle_path()?,
        recipient_compliance_position: cursor.read_u64()?,
        recipient_slot_id: cursor.read_fixed::<32>()?,
        recipient_slot_derivation: cursor.read_fixed::<32>()?,
        recipient_d: cursor.read_fixed::<32>()?,
        recipient_diversified_generator_affine: cursor.read_point_affine()?,
        recipient_transmission_key_affine: cursor.read_point_affine()?,
    })
}

fn encode_change_output(buf: &mut Vec<u8>, output: &TransferChangeOutputWitnessV18) {
    put_bytes(buf, &output.note_commitment);
    put_bytes(buf, &output.created_note_blinding);
    put_bytes(buf, &output.created_note_amount);
}

fn decode_change_output(cursor: &mut BinaryCursor<'_>) -> Result<TransferChangeOutputWitnessV18> {
    Ok(TransferChangeOutputWitnessV18 {
        note_commitment: cursor.read_fixed::<32>()?,
        created_note_blinding: cursor.read_fixed::<32>()?,
        created_note_amount: cursor.read_fixed::<32>()?,
    })
}

File: tools/gnark/internal/abi/transfer_witness_binary.go
package abi

import (
	"bytes"
	"fmt"

	"github.com/mizufinance/shieldd/tools/gnark/internal/generated"
)

const transferWitnessMagic = "PTWG"

type TransferComplianceCiphertextWitnessBinary struct {
	C2         [32]byte
	Ciphertext [][32]byte
	EPKAffine  PointAffineBinary
}

type TransferComplianceMetadataWitnessBinary struct {
	RingIDHash      [32]byte
	PolicyIDHash    [32]byte
	ResourceHash    [32]byte
	PermissionHash  [32]byte
	TargetTimestamp [32]byte
	SenderCoreSalt  [32]byte
	SenderExtSalt   [32]byte
	OutputCoreSalt  [32]byte
	OutputExtSalt   [32]byte
}

type TransferRequiredSpendWitnessBinary struct {
	Nullifier               [32]byte
	SpentNoteBlinding       [32]byte
	SpentNoteAmount         [32]byte
	SpentNoteAssetID        [32]byte
	StateCommitmentPosition uint64
	StateCommitmentAuthPath [][3][32]byte
	SpendAuthRandomizer     [32]byte
	RKAffine                PointAffineBinary
	HistoryRequired         bool
}

type TransferOptionalSpendWitnessBinary struct {
	Nullifier               [32]byte
	SpentNoteBlinding       [32]byte
	SpentNoteAmount         [32]byte
	StateCommitmentPosition uint64
	StateCommitmentAuthPath [][3][32]byte
	SpendAuthRandomizer     [32]byte
	RKAffine                PointAffineBinary
	IsDummy                 bool
	DummyNullifierSeed      [32]byte
	HistoryRequired         bool
}

type TransferReceiverOutputWitnessBinary struct {
	NoteCommitment                [32]byte
	CreatedNoteBlinding           [32]byte
	CreatedNoteAmount             [32]byte
	RecipientCompliancePath       MerklePathBinary
	RecipientCompliancePosition   uint64
	RecipientD                    [32]byte
	RecipientStatus               [32]byte
	RecipientDiversifiedGenerator PointAffineBinary
	RecipientTransmissionKey      PointAffineBinary
}

type TransferChangeOutputWitnessBinary struct {
	NoteCommitment      [32]byte
	CreatedNoteBlinding [32]byte
	CreatedNoteAmount   [32]byte
}

type TransferVolumeAccumulatorWitnessBinary struct {
	Nullifier         [32]byte
	Commitment        [32]byte
	DayStart          [32]byte
	ProofContext      [32]byte
	UseReal           bool
	StartsNewDay      bool
	Subject           [32]byte
	PriorVolume       [32]byte
	PriorBlinding     [32]byte
	PriorCommitment   [32]byte
	PriorPosition     uint64
	PriorAuthPath     [][3][32]byte
	SuccessorVolume   [32]byte
	SuccessorBlinding [32]byte
}

type TransferWitnessBinary struct {
	TotalLength uint32

	Anchor                [32]byte
	AssetAnchor           [32]byte
	ComplianceAnchor      [32]byte
	TargetTimestamp       [32]byte
	ClaimedStatementHash  [32]byte
	RoutingTags           [2][32]byte
	RoutingParameterSetID [32]byte
	RecentPositionFloor   [32]byte
	VolumeAccumulator     TransferVolumeAccumulatorWitnessBinary

	ActionBalanceBlinding    [32]byte
	NK                       [32]byte
	AssetPath                MerklePathBinary
	AssetPosition            uint64
	AssetIndexedLeaf         IndexedLeafBinary
	IsRegulated              bool
	RegulatedPrecision       uint8
	UnregulatedPrecision     uint8
	RoutingAsOfHeight        uint64
	SenderCompliancePath     MerklePathBinary
	SenderCompliancePosition uint64
	SenderD                  [32]byte
	SenderStatus             [32]byte
	TransferNonceRoot        [32]byte

	DetectionCiphertext [][32]byte
	Metadata            TransferComplianceMetadataWitnessBinary
	SenderCore          TransferComplianceCiphertextWitnessBinary
	SenderExt           TransferComplianceCiphertextWitnessBinary
	OutputCore          TransferComplianceCiphertextWitnessBinary
	OutputExt           TransferComplianceCiphertextWitnessBinary
	SenderRCore         [32]byte
	SenderRExt          [32]byte
	OutputRCore         [32]byte
	OutputRExt          [32]byte

	RequiredSpend  TransferRequiredSpendWitnessBinary
	OptionalSpend  TransferOptionalSpendWitnessBinary
	ReceiverOutput TransferReceiverOutputWitnessBinary
	ChangeOutput   TransferChangeOutputWitnessBinary

	AKAffine                   PointAffineBinary
	AssetIndexedLeafDKPub      PointAffineBinary
	AssetIndexedLeafRingPK     PointAffineBinary
	SenderDiversifiedGenerator PointAffineBinary
	SenderTransmissionKey      PointAffineBinary
}

func DecodeTransferWitness(
	payload []byte,
) (*TransferWitnessBinary, generated.TransferFamilySpec, error) {
	family, ok := generated.TransferFamilyByLabel("transfer")
	if !ok {
		return nil, generated.TransferFamilySpec{}, fmt.Errorf("missing generated transfer spec")
	}
	witness, err := decodeTransferWitness("Transfer", payload)
	if err != nil {
		return nil, generated.TransferFamilySpec{}, err
	}
	return witness, family, nil
}

func decodeTransferWitness(
	label string,
	payload []byte,
) (*TransferWitnessBinary, error) {
	reader := bytes.NewReader(payload)

	magic, err := readExact(reader, 4)
	if err != nil {
		return nil, err
	}
	if string(magic) != transferWitnessMagic {
		return nil, fmt.Errorf("invalid %sWitness magic %q", label, string(magic))
	}
	totalLength, err := readU32(reader)
	if err != nil {
		return nil, err
	}
	if totalLength != uint32(len(payload)) {
		return nil, fmt.Errorf("payload length mismatch: header=%d actual=%d", totalLength, len(payload))
	}

	witness := &TransferWitnessBinary{TotalLength: totalLength}
	if witness.Anchor, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.AssetAnchor, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.ComplianceAnchor, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.TargetTimestamp, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.ClaimedStatementHash, err = read32(reader); err != nil {
		return nil, err
	}
	for i := range witness.RoutingTags {
		if witness.RoutingTags[i], err = read32(reader); err != nil {
			return nil, err
		}
	}
	if witness.RoutingParameterSetID, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.RecentPositionFloor, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.VolumeAccumulator, err = readTransferVolumeAccumulator(reader); err != nil {
		return nil, fmt.Errorf("decode transfer volume accumulator: %w", err)
	}
	if witness.ActionBalanceBlinding, err = readFr32(reader); err != nil {
		return nil, err
	}
	if witness.NK, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.AssetPath, err = readMerklePath(reader); err != nil {
		return nil, err
	}
	if witness.AssetPosition, err = readU64(reader); err != nil {
		return nil, err
	}
	if witness.AssetIndexedLeaf, err = readIndexedLeaf(reader); err != nil {
		return nil, err
	}
	if witness.IsRegulated, err = readBool(reader); err != nil {
		return nil, err
	}
	if witness.RegulatedPrecision, err = readU8(reader); err != nil {
		return nil, err
	}
	if witness.UnregulatedPrecision, err = readU8(reader); err != nil {
		return nil, err
	}
	if witness.RoutingAsOfHeight, err = readU64(reader); err != nil {
		return nil, err
	}
	if witness.SenderCompliancePath, err = readMerklePath(reader); err != nil {
		return nil, err
	}
	if witness.SenderCompliancePosition, err = readU64(reader); err != nil {
		return nil, err
	}
	if witness.SenderD, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.SenderStatus, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.TransferNonceRoot, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.DetectionCiphertext, err = readVec32(reader); err != nil {
		return nil, err
	}
	if witness.Metadata, err = readTransferComplianceMetadata(reader); err != nil {
		return nil, err
	}
	if witness.SenderCore, err = readTransferComplianceTier(reader); err != nil {
		return nil, err
	}
	if witness.SenderExt, err = readTransferComplianceTier(reader); err != nil {
		return nil, err
	}
	if witness.OutputCore, err = readTransferComplianceTier(reader); err != nil {
		return nil, err
	}
	if witness.OutputExt, err = readTransferComplianceTier(reader); err != nil {
		return nil, err
	}
	if witness.SenderRCore, err = readFr32(reader); err != nil {
		return nil, err
	}
	if witness.SenderRExt, err = readFr32(reader); err != nil {
		return nil, err
	}
	if witness.OutputRCore, err = readFr32(reader); err != nil {
		return nil, err
	}
	if witness.OutputRExt, err = readFr32(reader); err != nil {
		return nil, err
	}

	if witness.RequiredSpend, err = readTransferRequiredSpend(reader); err != nil {
		return nil, fmt.Errorf("decode required transfer spend: %w", err)
	}
	if witness.OptionalSpend, err = readTransferOptionalSpend(reader); err != nil {
		return nil, fmt.Errorf("decode optional transfer spend: %w", err)
	}
	if witness.ReceiverOutput, err = readTransferReceiverOutput(reader); err != nil {
		return nil, fmt.Errorf("decode transfer receiver output: %w", err)
	}
	if witness.ChangeOutput, err = readTransferChangeOutput(reader); err != nil {
		return nil, fmt.Errorf("decode transfer change output: %w", err)
	}

	if witness.AKAffine, err = readPointAffine(reader); err != nil {
		return nil, err
	}
	if witness.AssetIndexedLeafDKPub, err = readPointAffine(reader); err != nil {
		return nil, err
	}
	if witness.AssetIndexedLeafRingPK, err = readPointAffine(reader); err != nil {
		return nil, err
	}
	if witness.SenderDiversifiedGenerator, err = readPointAffine(reader); err != nil {
		return nil, err
	}
	if witness.SenderTransmissionKey, err = readPointAffine(reader); err != nil {
		return nil, err
	}

	if extra, err := readExact(reader, int(reader.Len())); err != nil {
		return nil, err
	} else if len(extra) != 0 {
		return nil, fmt.Errorf("%s witness has %d trailing bytes", label, len(extra))
	}

	return witness, nil
}

func readTransferVolumeAccumulator(
	reader *bytes.Reader,
) (TransferVolumeAccumulatorWitnessBinary, error) {
	var value TransferVolumeAccumulatorWitnessBinary
	var err error
	if value.Nullifier, err = read32(reader); err != nil {
		return value, err
	}
	if value.Commitment, err = read32(reader); err != nil {
		return value, err
	}
	if value.DayStart, err = read32(reader); err != nil {
		return value, err
	}
	if value.ProofContext, err = read32(reader); err != nil {
		return value, err
	}
	if value.UseReal, err = readBool(reader); err != nil {
		return value, err
	}
	if value.StartsNewDay, err = readBool(reader); err != nil {
		return value, err
	}
	if value.Subject, err = read32(reader); err != nil {
		return value, err
	}
	if value.PriorVolume, err = read32(reader); err != nil {
		return value, err
	}
	if value.PriorBlinding, err = read32(reader); err != nil {
		return value, err
	}
	if value.PriorCommitment, err = read32(reader); err != nil {
		return value, err
	}
	if value.PriorPosition, err = readU64(reader); err != nil {
		return value, err
	}
	if value.PriorAuthPath, err = readTriplePath(reader); err != nil {
		return value, err
	}
	if value.SuccessorVolume, err = read32(reader); err != nil {
		return value, err
	}
	if value.SuccessorBlinding, err = read32(reader); err != nil {
		return value, err
	}
	return value, nil
}

func readTransferRequiredSpend(
	reader *bytes.Reader,
) (TransferRequiredSpendWitnessBinary, error) {
	var spend TransferRequiredSpendWitnessBinary
	var err error
	if spend.Nullifier, err = read32(reader); err != nil {
		return spend, err
	}
	if spend.SpentNoteBlinding, err = read32(reader); err != nil {
		return spend, err
	}
	if spend.SpentNoteAmount, err = read32(reader); err != nil {
		return spend, err
	}
	if spend.SpentNoteAssetID, err = read32(reader); err != nil {
		return spend, err
	}
	if spend.StateCommitmentPosition, err = readU64(reader); err != nil {
		return spend, err
	}
	if spend.StateCommitmentAuthPath, err = readTriplePath(reader); err != nil {
		return spend, err
	}
	if spend.SpendAuthRandomizer, err = readFr32(reader); err != nil {
		return spend, err
	}
	if spend.RKAffine, err = readPointAffine(reader); err != nil {
		return spend, err
	}
	if spend.HistoryRequired, err = readBool(reader); err != nil {
		return spend, err
	}
	return spend, nil
}

func readTransferOptionalSpend(
	reader *bytes.Reader,
) (TransferOptionalSpendWitnessBinary, error) {
	var spend TransferOptionalSpendWitnessBinary
	var err error
	if spend.Nullifier, err = read32(reader); err != nil {
		return spend, err
	}
	if spend.SpentNoteBlinding, err = read32(reader); err != nil {
		return spend, err
	}
	if spend.SpentNoteAmount, err = read32(reader); err != nil {
		return spend, err
	}
	if spend.StateCommitmentPosition, err = readU64(reader); err != nil {
		return spend, err
	}
	if spend.StateCommitmentAuthPath, err = readTriplePath(reader); err != nil {
		return spend, err
	}
	if spend.SpendAuthRandomizer, err = readFr32(reader); err != nil {
		return spend, err
	}
	if spend.RKAffine, err = readPointAffine(reader); err != nil {
		return spend, err
	}
	if spend.IsDummy, err = readBool(reader); err != nil {
		return spend, err
	}
	if spend.DummyNullifierSeed, err = read32(reader); err != nil {
		return spend, err
	}
	if spend.HistoryRequired, err = readBool(reader); err != nil {
		return spend, err
	}
	return spend, nil
}

func readTransferReceiverOutput(
	reader *bytes.Reader,
) (TransferReceiverOutputWitnessBinary, error) {
	var output TransferReceiverOutputWitnessBinary
	var err error
	if output.NoteCommitment, err = read32(reader); err != nil {
		return output, err
	}
	if output.CreatedNoteBlinding, err = read32(reader); err != nil {
		return output, err
	}
	if output.CreatedNoteAmount, err = read32(reader); err != nil {
		return output, err
	}
	if output.RecipientCompliancePath, err = readMerklePath(reader); err != nil {
		return output, err
	}
	if output.RecipientCompliancePosition, err = readU64(reader); err != nil {
		return output, err
	}
	if output.RecipientD, err = read32(reader); err != nil {
		return output, err
	}
	if output.RecipientStatus, err = read32(reader); err != nil {
		return output, err
	}
	if output.RecipientDiversifiedGenerator, err = readPointAffine(reader); err != nil {
		return output, err
	}
	if output.RecipientTransmissionKey, err = readPointAffine(reader); err != nil {
		return output, err
	}
	return output, nil
}

func readTransferChangeOutput(
	reader *bytes.Reader,
) (TransferChangeOutputWitnessBinary, error) {
	var output TransferChangeOutputWitnessBinary
	var err error
	if output.NoteCommitment, err = read32(reader); err != nil {
		return output, err
	}
	if output.CreatedNoteBlinding, err = read32(reader); err != nil {
		return output, err
	}
	if output.CreatedNoteAmount, err = read32(reader); err != nil {
		return output, err
	}
	return output, nil
}

func readTransferComplianceTier(
	reader *bytes.Reader,
) (TransferComplianceCiphertextWitnessBinary, error) {
	var tier TransferComplianceCiphertextWitnessBinary
	var err error
	if tier.C2, err = read32(reader); err != nil {
		return tier, err
	}
	if tier.Ciphertext, err = readVec32(reader); err != nil {
		return tier, err
	}
	if tier.EPKAffine, err = readPointAffine(reader); err != nil {
		return tier, err
	}
	return tier, nil
}

func readTransferComplianceMetadata(
	reader *bytes.Reader,
) (TransferComplianceMetadataWitnessBinary, error) {
	var metadata TransferComplianceMetadataWitnessBinary
	var err error
	if metadata.RingIDHash, err = read32(reader); err != nil {
		return metadata, err
	}
	if metadata.PolicyIDHash, err = read32(reader); err != nil {
		return metadata, err
	}
	if metadata.ResourceHash, err = read32(reader); err != nil {
		return metadata, err
	}
	if metadata.PermissionHash, err = read32(reader); err != nil {
		return metadata, err
	}
	if metadata.TargetTimestamp, err = read32(reader); err != nil {
		return metadata, err
	}
	if metadata.SenderCoreSalt, err = read32(reader); err != nil {
		return metadata, err
	}
	if metadata.SenderExtSalt, err = read32(reader); err != nil {
		return metadata, err
	}
	if metadata.OutputCoreSalt, err = read32(reader); err != nil {
		return metadata, err
	}
	if metadata.OutputExtSalt, err = read32(reader); err != nil {
		return metadata, err
	}
	return metadata, nil
}
